graph/connected_components_with_bfs.cpp

```diff
@@ -0,0 +1,92 @@
+ *
+ * \author [Ritik Rathor](https://github.com/ritik-369)
+ *
+ * \details
+ * A graph is a collection of nodes also called vertices and these vertices
+ * are connected by edges. A connected component in a graph refers to a set of
+ * vertices which are reachable form one another.
+ *
+ * <pre>
+ * Example - Here is graph with 3 connected components
+ *
+ *      1   4           5               8
+ *     / \ /           / \             / \
+ *    2---3           6   7           9   10
+ *
+ *    first          second           third
+ *    component      component        component
+ * </pre>
+ *
+ */
+ 
+
+//                  WE WILL LEARN CALCULATING NO. OF CONNECTED COMPONENTS IN GRAPH USING BFS BY SOLVING A HACKEREARTH QUESTION
+//                  LINK=   https://www.hackerearth.com/problem/algorithm/connected-components-in-a-graph/
+
+
+#include <iostream>  // for IO operations
+#include <vector>    //for using vectors to make graph or adjacency list
+#include <queue>     //for using queue data structure for pushing and poping nodes during bfs
+
+/**
+ * @brief connected_components_bfs function
+ * @returns number of connected components on exit
+ */
+int connected_components_bfs(std::vector<std::vector<int>>adj,int nodes) // {adj} is adjacency list of graph and {nodes} is number of nodes in graph 
+{
+   int components=0;          // initialize no. of components in graph as 0
+   int start=1;              // source node declaration as number 1 we can assign 0 if number of nodes starts from zero
+   int end=(nodes+start-1);
+   std::vector<bool>vis(nodes+1,0);  // vector for visited node used during running of bfs
+   std::queue<int>q;                  // {q} parameter is queue which stores childs to be visited in bfs
+   for(int node=start;node<=end;node++)          // running loop from starting node to end node
+   {
+      if(vis[node])continue;      // if node is visted means its component in graph has been calculted so we continue loop
+      else components++;          // else we increase the no. of components by one and move further    
+      int cur=node;               // {cur} is current node which is processing in bfs and we initilize it as node which is unvisited
+      q.push(cur);                  //pushing into queue for bfs
+      vis[cur]=1;                    // mark node visited so we will not process it later
+      while(!q.empty())             // running bfs loop
+      {
+         cur=q.front();            // extracting data of front node from queue
+         q.pop();                  // removing front node
+         for(auto child:adj[cur]) if(!vis[child])    //{child} is child of current node which is processing under bfs and check if child is not visited
+         {
+            q.push(child);               // pushing non visited node in queue
+            vis[child]=1;                // marking it as visited
+         }
+      }
+   }
+   return components;                   // returning components in graph
+}
+
```

Comment:
Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃